### PR TITLE
bugfix for guard_view and zip_view bracket operator 

### PR DESCRIPTION
**Pull request by danhoeflinger**
The guard_view and zip_view bracket operator was previously using signed int32 for the type of the index.  

The fix changes this index to be a template argument which is in line with other views in this file.  

I've added some tests to cover the overflow case for signed int32 for zip_view and guard_view.  They fail without the changes, and pass after the changes.

This PR impacts https://github.com/oneapi-src/oneDPL/pull/608 ,  guard_view's bracket operator is used in the by_segment routines and this bug impacts correctness of those algorithms as they are written.

### CODE REVIEW

**Review by julianmi — APPROVED** [review]
LGTM. Maybe wait for another opinion before merging.

**Comment by akukanov** [comment]
According to the C++ standard, the parameter to `zip_view::operator[]` should be of the difference type for the underlying range. In the long term, I think this is what we should use for all our view implementations. In the short term, the changes in the library headers look fine to me (and I have not reviewed the tests). But let's perhaps add a comment about the eventual need to align with the standard on that.

**Comment by danhoeflinger** [comment]
> @akukanov According to the C++ standard, the parameter to `zip_view::operator[]` should be of the difference type for the underlying range. In the long term, I think this is what we should use for all our view implementations. In the short term, the changes in the library headers look fine to me (and I have not reviewed the tests). But let's perhaps add a comment about the eventual need to align with the standard on that.

Yes, this was my first fix locally, but since other views in this file used this template strategy, I followed suit.  I can add a comment note to switch in the future.  Thanks for taking a look.

File: test/parallel_api/ranges/zip_view.pass.cpp
Comment (akukanov) [inline test/parallel_api/ranges/zip_view.pass.cpp:46]: Allocating 4GB of memory twice is quite a lot, and it may take quite some time as well since all the elements should be initialized. I would recommend an approach that requires less space, for example, using views::iota and views::transform to map big indexes to a reasonably small data container.

Comment (danhoeflinger) [reply test/parallel_api/ranges/zip_view.pass.cpp]: I adjusted the test to reuse existing small vectors as you describe.

**Review by julianmi — DISMISSED** [review]

**Review by julianmi — APPROVED** [review]

**Pull request merged**
